README.md
[ENGLISH](./README_EN.md)

<h1 align="center">Color Extension</h1>

<p align="center">
  <a href="https://travis-ci.com/DataV-Team/color">
    <img src="https://img.shields.io/travis/com/DataV-Team/color.svg" alt="Travis CI" />
  </a>
  <a href="https://github.com/DataV-Team/color/blob/master/LICENSE">
    <img src="https://img.shields.io/github/license/DataV-Team/color.svg" alt="LICENSE" />
  </a>
  <a href="https://www.npmjs.com/package/@jiaminghi/color">
    <img src="https://img.shields.io/npm/v/@jiaminghi/color.svg" alt="NPM" />
  </a>
</p>

### 这是一个提供如下几个方法的颜色插件。

- **[isHex](#isHex)**

  判断一个颜色是否为 Hex 类型

- **[isRgb](#isRgb)**

  判断一个颜色是否为 Rgb 类型

- **[isRgba](#isRgba)**

  判断一个颜色是否为 Rgba 类型

- **[isRgbOrRgba](#isRgbOrRgba)**

  判断一个颜色是否为 Rgb 或 Rgba 类型

- **[darken](#darken)**

  加深颜色

- **[lighten](#lighten)**

  提亮颜色

- **[fade](#fade)**

  调节颜色透明度

- **[toHex](#toHex)**

  转换成 Hex 颜色

- **[toRgb](#toRgb)**

  将颜色转为 Rgb 或 Rgba 颜色

- **[getOpacity](#getOpacity)**

  获取颜色透明度

- **[getRgbValue](#getRgbValue)**

  获取颜色的 Rgb 值

- **[getRgbaValue](#getRgbaValue)**

  获取颜色的 Rgba 值

- **[getColorFromRgbValue](#getColorFromRgbValue)**

  从 Rgb 或 Rgba 值获取颜色

### 附录

- **[颜色关键字](#颜色关键字)**

  内置颜色关键字

### npm 安装

```shell
$ npm install @jiaminghi/color
```

### 使用

```javascript
import { toHex } from '@jiaminghi/color'

// do something
```

### 快速体验

```html
<!--调试版-->
<script src="https://unpkg.com/@jiaminghi/color/dist/index.js"></script>
<!--压缩版-->
<script src="https://unpkg.com/@jiaminghi/color/dist/index.min.js"></script>
<script>
  const { darken, lighten } = window.Color
  // do something
</script>
```

---

<h3 align="center">示例</h3>

#### isHex

```typescript
/**
 * @description 判断一个颜色是否为hex类型
 * @param {string} color 可能是hex颜色的字符串
 * @return {boolean} 判断结果
 */
type isHex = (color: string) => boolean

isHex('#000') // true
isHex('#333333') // true
isHex('Not A Color') // false
isHex('rgb(0,0,0)') // false
isHex('rgba(0,0,0,1)') // false
```

#### isRgb

```typescript
/**
 * @description 判断一个颜色是否为rgb类型
 * @param {string} color 可能是rgb颜色的字符串
 * @return {boolean} 判断结果
 */
type isRgb = (color: string) => boolean

isRgb('rgb(0,0,0)') // true
isRgb('RGB(0,0,0)') // true
isRgb('Not A Color') // false
isRgb('#000') // false
isRgb('#000000') // false
```

#### isRgba

```typescript
/**
 * @description 判断一个颜色是否为rgba类型
 * @param {string} color 可能是rgba颜色的字符串
 * @return {boolean} 判断结果
 */
type isRgba = (color: string) => boolean

isRgba('rgba(0,0,0,1)') // true
isRgba('rgb(0,0,0)') // false
isRgba('Not A Color') // false
isRgba('#000') // false
isRgba('#000000') // false
```

#### isRgbOrRgba

```typescript
/**
 * @description 判断一个颜色是否为rgb或rgba类型
 * @param {string} color 可能是rgb或rgba颜色的字符串
 * @return {boolean} 判断结果
 */
type isRgbOrRgba = (color: string) => boolean

isRgbOrRgba('rgb(0,0,0)') // true
isRgbOrRgba('RGB(0,0,0)') // true
isRgbOrRgba('rgba(0,0,0,1)') // true
isRgbOrRgba('#000') // false
isRgbOrRgba('Not A Color') // false
```

#### darken

```typescript
/**
 * @description 加深颜色
 * @param {string} color   Hex|Rgb|Rgba颜色或颜色关键字
 * @param {number} percent 加深的百分比 (1-100)
 * @return {string|null} Rgba颜色 (无效输入将返回null)
 */
type darken = (color: string, percent: number) => string | null

const before = '#3080E8'

const after = darken(color, 20)
// after = 'rgba(0,77,181,1)'
```

<p align="center">
  <img width="180px" src="./exampleImgs/1.jpg" />
</p>

#### lighten

```typescript
/**
 * @description 提亮颜色
 * @param {string} color   Hex|Rgb|Rgba颜色或颜色关键字
 * @param {number} percent 提亮的百分比 (1-100)
 * @return {string|null} Rgba颜色 (无效输入将返回null)
 */
type lighten = (color: string, percent: number) => string | null

const before = '#3080E8'

const after = lighten(color, 20)
// after = 'rgba(99,179,255,1)'
```

<p align="center">
    <img width="180px" src="./exampleImgs/2.jpg" />
</p>

#### fade

```typescript
/**
 * @description 调节颜色透明度
 * @param {string} color   Hex|Rgb|Rgba颜色或颜色关键字
 * @param {number} Percent 透明度百分比
 * @return {string|null} Rgba颜色 (无效输入将返回null)
 */
type fade = (color: string, percent: number) => string | null

const before = '#3080E8'

const after = lighten(color, 20)
// after = 'rgba(48,128,232,0.2)'
```

<p align="center">
  <img width="180px" src="./exampleImgs/3.jpg" />
</p>

#### toHex

```typescript
/**
 * @description 转换成Hex颜色
 * @param {string} color Hex|Rgb|Rgba颜色或颜色关键字
 * @return {string|null} Hex颜色 (无效输入将返回null)
 */
type toHex = (color: string) => string | false

const before = 'rgb(48,128,232)'

const after = toHex(before)
// after = '#3080e8'
```

#### toRgb

```typescript
/**
 * @description 将颜色转为Rgb或Rgba颜色
 * @param {string} color   Hex|Rgb|Rgba颜色或颜色关键字
 * @param {number} opacity 颜色的透明度 (输入该参数将生成Rgba颜色)
 * @return {string|null} Rgb或Rgba颜色 (无效输入将返回null)
 */
type toRgb = (color: string, opacity: number) => string | null

const before = '#3080E8'

const after1 = toRgb(before)
// after1 = 'rgb(48,128,232)'
const after2 = toRgb(before, 0.2)
// after2 = 'rgba(48,128,232,0.2)'
```

#### getOpacity

```typescript
/**
 * @description 获取颜色透明度
 * @param {string} color Hex|Rgb|Rgba颜色或颜色关键字
 * @return {number|null} 颜色透明度 (无效输入将返回null)
 */
type getOpacity = (color: string) => number | null

const color1 = '#3080E8'
const color2 = 'rgba(48,128,232,0.2)'

const opacity1 = getOpacity(color1)
// opacity1 = 1
const opacity2 = getOpacity(color2)
// opacity2 = 0.2
```

#### getRgbValue

```typescript
/**
 * @description 获取颜色的Rgb值
 * @param {string} color Hex|Rgb|Rgba颜色或颜色关键字
 * @return {RgbValue|null} Rgb值 (无效输入将返回null)
 */
type RgbValue = [number, number, number]
type getRgbValue = (color: string) => RgbValue | null

const color = '#3080E8'

const rgbValue = getRgbValue(color)
// rgbValue = [48, 128, 232]
```

#### getRgbaValue

```typescript
/**
 * @description 获取颜色的Rgba值
 * @param {String} color Hex|Rgb|Rgba颜色或颜色关键字
 * @return {RgbaValue|null} Rgba值 (无效输入将返回null)
 */
type RgbaValue = [number, number, number, number]
type getRgbaValue = (color: string) => RgbaValue

const color1 = '#3080E8'
const color2 = 'rgba(48,128,232,0.2)'

const rgbaValue1 = getRgbaValue(color1)
// rgbaValue1 = [48, 128, 232, 1]
const rgbaValue2 = getRgbaValue(color2)
// rgbaValue2 = [48, 128, 232, 0.2]
```

#### getColorFromRgbValue

```typescript
/**
 * @description 从Rgb或Rgba值获取颜色
 * @param {Value} value Rgb或Rgba颜色的值
 * @return {string|null} Rgb颜色或Rgba颜色 (无效输入将返回null)
 */
type RgbValue = [number, number, number]
type RgbaValue = [number, number, number, number]
type Value = RgbValue | RgbaValue
type getColorFromRgbValue = (value: Value) => string | null

const value1 = [48, 128, 232]
const value2 = [48, 128, 232, 0.2]

const color1 = getColorFromRgbValue(value1)
// color1 = 'rgb(48,128,232)'
const color2 = getColorFromRgbValue(value2)
// color2 = 'rgba(48,128,232,0.2)'
```

#### 颜色关键字

<p align="center">
  <img width="750px" src="./exampleImgs/keywords.png" />
</p>


CHANGELOG.md
# 1.1.3 (2023-01-02)

### Bug Fixes

- **chore:** Module export entry error

# 1.1.2 (2023-01-02)

### Perfect

- **chore:** Remove default export.

# 1.1.1 (2020-05-14)

### Bug Fixes

- **types:** Absolute path cause import error.

# 1.1.0 (2020-04-21)

### Perfect

- **devDep:** Add rollup into devDep.

# 1.0.0 (2020-04-11)

### Upgrade

- **typescript:** Rewrite with typescript.

### New API

- **isHex:** Test is a Hex color.
- **isRgb:** Test is a Rgb color.
- **isRgba:** Test is a Rgba color.
- **isRgbOrRgba:** Test is a Rgb or Rgba color.

# 0.1.1-alpha (2019-08-29)

### Perfect

- **babel:** Upgrade babel compilation mode.

# 0.1.0-alpha (2019-07-07)

### Perfect

- **keywords:** Support color keywords like red, blue, transparent and etc.

# 0.0.6-alpha (2019-06-13)

### Bug Fixes

- **getOpacity:** Fix the exception caused by the decimal in the rgba color (Like 'rgba(123.3333, 111, 111, 0.323)').

# 0.0.5-alpha (2019-06-06)

### Perfect

- **ES5:** Use babel to transcode to `ES5`.

# 0.0.4-alpha (2019-05-15)

### Publish

- **version:** V 0.0.4-alpha.

# 0.0.3-alpha (2019-04-30)

### Bug Fixes

- **getOpacity:** Abnormal calculation due to spaces.

# 0.0.2-alpha (2019-04-08)

### Publish

- **version:** V 0.0.2-alpha.

# 0.0.1-alpha (2019-04-04)

### Publish

- **version:** V 0.0.1-alpha.


lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

function __spreadArrays() {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
}

var ColorKeywords;

(function (ColorKeywords) {
  ColorKeywords["transparent"] = "rgba(0,0,0,0)";
  ColorKeywords["black"] = "#000000";
  ColorKeywords["silver"] = "#C0C0C0";
  ColorKeywords["gray"] = "#808080";
  ColorKeywords["white"] = "#FFFFFF";
  ColorKeywords["maroon"] = "#800000";
  ColorKeywords["red"] = "#FF0000";
  ColorKeywords["purple"] = "#800080";
  ColorKeywords["fuchsia"] = "#FF00FF";
  ColorKeywords["green"] = "#008000";
  ColorKeywords["lime"] = "#00FF00";
  ColorKeywords["olive"] = "#808000";
  ColorKeywords["yellow"] = "#FFFF00";
  ColorKeywords["navy"] = "#000080";
  ColorKeywords["blue"] = "#0000FF";
  ColorKeywords["teal"] = "#008080";
  ColorKeywords["aqua"] = "#00FFFF";
  ColorKeywords["aliceblue"] = "#f0f8ff";
  ColorKeywords["antiquewhite"] = "#faebd7";
  ColorKeywords["aquamarine"] = "#7fffd4";
  ColorKeywords["azure"] = "#f0ffff";
  ColorKeywords["beige"] = "#f5f5dc";
  ColorKeywords["bisque"] = "#ffe4c4";
  ColorKeywords["blanchedalmond"] = "#ffebcd";
  ColorKeywords["blueviolet"] = "#8a2be2";
  ColorKeywords["brown"] = "#a52a2a";
  ColorKeywords["burlywood"] = "#deb887";
  ColorKeywords["cadetblue"] = "#5f9ea0";
  ColorKeywords["chartreuse"] = "#7fff00";
  ColorKeywords["chocolate"] = "#d2691e";
  ColorKeywords["coral"] = "#ff7f50";
  ColorKeywords["cornflowerblue"] = "#6495ed";
  ColorKeywords["cornsilk"] = "#fff8dc";
  ColorKeywords["crimson"] = "#dc143c";
  ColorKeywords["cyan"] = "#00ffff";
  ColorKeywords["darkblue"] = "#00008b";
  ColorKeywords["darkcyan"] = "#008b8b";
  ColorKeywords["darkgoldenrod"] = "#b8860b";
  ColorKeywords["darkgray"] = "#a9a9a9";
  ColorKeywords["darkgreen"] = "#006400";
  ColorKeywords["darkgrey"] = "#a9a9a9";
  ColorKeywords["darkkhaki"] = "#bdb76b";
  ColorKeywords["darkmagenta"] = "#8b008b";
  ColorKeywords["darkolivegreen"] = "#556b2f";
  ColorKeywords["darkorange"] = "#ff8c00";
  ColorKeywords["darkorchid"] = "#9932cc";
  ColorKeywords["darkred"] = "#8b0000";
  ColorKeywords["darksalmon"] = "#e9967a";
  ColorKeywords["darkseagreen"] = "#8fbc8f";
  ColorKeywords["darkslateblue"] = "#483d8b";
  ColorKeywords["darkslategray"] = "#2f4f4f";
  ColorKeywords["darkslategrey"] = "#2f4f4f";
  ColorKeywords["darkturquoise"] = "#00ced1";
  ColorKeywords["darkviolet"] = "#9400d3";
  ColorKeywords["deeppink"] = "#ff1493";
  ColorKeywords["deepskyblue"] = "#00bfff";
  ColorKeywords["dimgray"] = "#696969";
  ColorKeywords["dimgrey"] = "#696969";
  ColorKeywords["dodgerblue"] = "#1e90ff";
  ColorKeywords["firebrick"] = "#b22222";
  ColorKeywords["floralwhite"] = "#fffaf0";
  ColorKeywords["forestgreen"] = "#228b22";
  ColorKeywords["gainsboro"] = "#dcdcdc";
  ColorKeywords["ghostwhite"] = "#f8f8ff";
  ColorKeywords["gold"] = "#ffd700";
  ColorKeywords["goldenrod"] = "#daa520";
  ColorKeywords["greenyellow"] = "#adff2f";
  ColorKeywords["grey"] = "#808080";
  ColorKeywords["honeydew"] = "#f0fff0";
  ColorKeywords["hotpink"] = "#ff69b4";
  ColorKeywords["indianred"] = "#cd5c5c";
  ColorKeywords["indigo"] = "#4b0082";
  ColorKeywords["ivory"] = "#fffff0";
  ColorKeywords["khaki"] = "#f0e68c";
  ColorKeywords["lavender"] = "#e6e6fa";
  ColorKeywords["lavenderblush"] = "#fff0f5";
  ColorKeywords["lawngreen"] = "#7cfc00";
  ColorKeywords["lemonchiffon"] = "#fffacd";
  ColorKeywords["lightblue"] = "#add8e6";
  ColorKeywords["lightcoral"] = "#f08080";
  ColorKeywords["lightcyan"] = "#e0ffff";
  ColorKeywords["lightgoldenrodyellow"] = "#fafad2";
  ColorKeywords["lightgray"] = "#d3d3d3";
  ColorKeywords["lightgreen"] = "#90ee90";
  ColorKeywords["lightgrey"] = "#d3d3d3";
  ColorKeywords["lightpink"] = "#ffb6c1";
  ColorKeywords["lightsalmon"] = "#ffa07a";
  ColorKeywords["lightseagreen"] = "#20b2aa";
  ColorKeywords["lightskyblue"] = "#87cefa";
  ColorKeywords["lightslategray"] = "#778899";
  ColorKeywords["lightslategrey"] = "#778899";
  ColorKeywords["lightsteelblue"] = "#b0c4de";
  ColorKeywords["lightyellow"] = "#ffffe0";
  ColorKeywords["limegreen"] = "#32cd32";
  ColorKeywords["linen"] = "#faf0e6";
  ColorKeywords["magenta"] = "#ff00ff";
  ColorKeywords["mediumaquamarine"] = "#66cdaa";
  ColorKeywords["mediumblue"] = "#0000cd";
  ColorKeywords["mediumorchid"] = "#ba55d3";
  ColorKeywords["mediumpurple"] = "#9370db";
  ColorKeywords["mediumseagreen"] = "#3cb371";
  ColorKeywords["mediumslateblue"] = "#7b68ee";
  ColorKeywords["mediumspringgreen"] = "#00fa9a";
  ColorKeywords["mediumturquoise"] = "#48d1cc";
  ColorKeywords["mediumvioletred"] = "#c71585";
  ColorKeywords["midnightblue"] = "#191970";
  ColorKeywords["mintcream"] = "#f5fffa";
  ColorKeywords["mistyrose"] = "#ffe4e1";
  ColorKeywords["moccasin"] = "#ffe4b5";
  ColorKeywords["navajowhite"] = "#ffdead";
  ColorKeywords["oldlace"] = "#fdf5e6";
  ColorKeywords["olivedrab"] = "#6b8e23";
  ColorKeywords["orange"] = "#ffa500";
  ColorKeywords["orangered"] = "#ff4500";
  ColorKeywords["orchid"] = "#da70d6";
  ColorKeywords["palegoldenrod"] = "#eee8aa";
  ColorKeywords["palegreen"] = "#98fb98";
  ColorKeywords["paleturquoise"] = "#afeeee";
  ColorKeywords["palevioletred"] = "#db7093";
  ColorKeywords["papayawhip"] = "#ffefd5";
  ColorKeywords["peachpuff"] = "#ffdab9";
  ColorKeywords["peru"] = "#cd853f";
  ColorKeywords["pink"] = "#ffc0cb";
  ColorKeywords["plum"] = "#dda0dd";
  ColorKeywords["powderblue"] = "#b0e0e6";
  ColorKeywords["rosybrown"] = "#bc8f8f";
  ColorKeywords["royalblue"] = "#4169e1";
  ColorKeywords["saddlebrown"] = "#8b4513";
  ColorKeywords["salmon"] = "#fa8072";
  ColorKeywords["sandybrown"] = "#f4a460";
  ColorKeywords["seagreen"] = "#2e8b57";
  ColorKeywords["seashell"] = "#fff5ee";
  ColorKeywords["sienna"] = "#a0522d";
  ColorKeywords["skyblue"] = "#87ceeb";
  ColorKeywords["slateblue"] = "#6a5acd";
  ColorKeywords["slategray"] = "#708090";
  ColorKeywords["snow"] = "#fffafa";
  ColorKeywords["springgreen"] = "#00ff7f";
  ColorKeywords["steelblue"] = "#4682b4";
  ColorKeywords["tan"] = "#d2b48c";
  ColorKeywords["thistle"] = "#d8bfd8";
  ColorKeywords["tomato"] = "#ff6347";
  ColorKeywords["turquoise"] = "#40e0d0";
  ColorKeywords["violet"] = "#ee82ee";
  ColorKeywords["wheat"] = "#f5deb3";
  ColorKeywords["whitesmoke"] = "#f5f5f5";
  ColorKeywords["yellowgreen"] = "#9acd32";
})(ColorKeywords || (ColorKeywords = {}));

var ColorKeywords$1 = ColorKeywords;

/**
 * @description Test if a hex color
 * @param {string} color color string
 * @return {boolean} Test result
 */

function isHex(color) {
  if (typeof color !== 'string') return false;
  color = color.toLowerCase();
  return /^#([0-9a-fA-f]{3}|[0-9a-fA-f]{6})$/.test(color);
}
/**
 * @description Test if a rgb color
 * @param {string} color color string
 * @return {boolean} Test result
 */

function isRgb(color) {
  if (typeof color !== 'string') return false;
  color = color.toLowerCase();
  return /^(rgb\(|RGB\()/.test(color);
}
/**
 * @description Test if a rgba color
 * @param {string} color color string
 * @return {boolean} Test result
 */

function isRgba(color) {
  if (typeof color !== 'string') return false;
  color = color.toLowerCase();
  return /^(rgba|RGBA)/.test(color);
}
/**
 * @description Test if a rgb or rgba color
 * @param {string} color color string
 * @return {boolean} Test result
 */

function isRgbOrRgba(color) {
  return /^(rgb|rgba|RGB|RGBA)/.test(color);
}
/**
 * @description Get color by keyword
 * @param {keyof typeof ColorKeywords} keyword Color keyword like red, green and etc.
 * @return {ColorKeywords} Hex or rgba color
 */

function getColorByKeyword(keyword) {
  return ColorKeywords$1[keyword];
}
/**
 * @description Color validator
 * @param {string} color Hex|Rgb|Rgba color or color keywords
 * @return {string} Color (Invalid input will throw an error)
 */


function validator(color) {
  if (isHex(color) || isRgbOrRgba(color)) return color;
  var keywordColor = getColorByKeyword(color);
  if (!keywordColor) throw new Error("Color: Invalid Input of " + color);
  return keywordColor;
}
/**
 * @description Get the rgb value of the hex color
 * @param {string} color Hex color
 * @return {RgbValue} Rgb value of the color
 */


function getRgbValueFromHex(color) {
  color = color.replace('#', '');
  if (color.length === 3) color = Array.from(color).map(function (hexNum) {
    return hexNum + hexNum;
  }).join('');
  var colorValues = color.split('');
  return new Array(3).fill(0).map(function (_, i) {
    return parseInt("0x" + colorValues[i * 2] + colorValues[i * 2 + 1]);
  });
}
/**
 * @description Get the rgb value of the rgb/rgba color
 * @param {string} color Hex color
 * @return {RgbValue} Rgb value of the color
 */


function getRgbValueFromRgb(color) {
  return color.replace(/rgb\(|rgba\(|\)/g, '').split(',').slice(0, 3).map(function (n) {
    return parseInt(n);
  });
}
/**
 * @description Get the Rgb value of the color
 * @param {string} color Hex|Rgb|Rgba color or color keyword
 * @return {RgbValue} Rgb value of the color
 */


function getRgbValue(color) {
  var validColor = validator(color);
  var lowerColor = validColor.toLowerCase();
  return isHex(lowerColor) ? getRgbValueFromHex(lowerColor) : getRgbValueFromRgb(lowerColor);
}
/**
 * @description Get the opacity of color
 * @param {string} color Hex|Rgb|Rgba color or color keyword
 * @return {number} Color opacity
 */

function getOpacity(color) {
  var validColor = validator(color);
  if (!isRgba(validColor)) return 1;
  return Number(validColor.toLowerCase().split(',').slice(-1)[0].replace(/[)|\s]/g, ''));
}
/**
 * @description Get the Rgba value of the color
 * @param {string} color Hex|Rgb|Rgba color or color keyword
 * @return {RgbaValue} Rgba value of the color
 */

function getRgbaValue(color) {
  var rgbValue = getRgbValue(color);
  return rgbValue && __spreadArrays(rgbValue, [getOpacity(color)]);
}
/**
 * @description Convert color to Rgb|Rgba color
 * @param {string} color   Hex|Rgb|Rgba color or color keyword
 * @param {number} opacity The opacity of color
 * @return {string} Rgb|Rgba color
 */

function toRgb(color, opacity) {
  var rgbValue = getRgbValue(color);
  return typeof opacity === 'number' ? "rgba(" + rgbValue.join(',') + "," + opacity + ")" : "rgb(" + rgbValue.join(',') + ")";
}
/**
 * @description Convert color to Hex color
 * @param {string} color Hex|Rgb|Rgba color or color keyword
 * @return {string} Hex color
 */

function toHex(color) {
  if (isHex(color)) return color;
  var colorValue = getRgbValue(color);

  var format10To16 = function format10To16(_) {
    return Number(_).toString(16).padStart(2, '0');
  };

  return "#" + colorValue.map(format10To16).join('');
}
/**
 * @description Get Color from Rgb|Rgba value
 * @param {RgbValue|RgbaValue} value Rgb|Rgba color value
 * @return {string} Rgb|Rgba color
 */

function getColorFromRgbValue(value) {
  if (!Array.isArray(value)) throw new Error("getColorFromRgbValue: " + value + " is not an array");
  var length = value.length;
  if (length !== 3 && length !== 4) throw new Error("getColorFromRgbValue: value length should be 3 or 4");
  return (length === 3 ? 'rgb(' : 'rgba(') + value.join(',') + ')';
}
/**
 * @description Deepen color
 * @param {string} color   Hex|Rgb|Rgba color or color keyword
 * @param {number} percent of Deepen (1-100)
 * @return {string} Rgba color
 */

function darken(color, percent) {
  if (percent === void 0) {
    percent = 0;
  }

  var rgbaValue = getRgbaValue(color);
  rgbaValue = rgbaValue.map(function (v, i) {
    return i === 3 ? v : v - Math.ceil(2.55 * percent);
  }).map(function (v) {
    return v < 0 ? 0 : v;
  });
  return getColorFromRgbValue(rgbaValue);
}
/**
 * @description Brighten color
 * @param {string} color   Hex|Rgb|Rgba color or color keyword
 * @param {number} percent of brighten (1-100)
 * @return {string} Rgba color
 */

function lighten(color, percent) {
  if (percent === void 0) {
    percent = 0;
  }

  var rgbaValue = getRgbaValue(color);
  rgbaValue = rgbaValue.map(function (v, i) {
    return i === 3 ? v : v + Math.ceil(2.55 * percent);
  }).map(function (v) {
    return v > 255 ? 255 : v;
  });
  return getColorFromRgbValue(rgbaValue);
}
/**
 * @description Adjust color opacity
 * @param {string} color   Hex|Rgb|Rgba color or color keyword
 * @param {number} percent of opacity
 * @return {string} Rgba color
 */

function fade(color, percent) {
  if (percent === void 0) {
    percent = 100;
  }

  var rgbValue = getRgbValue(color);
  return getColorFromRgbValue(__spreadArrays(rgbValue, [percent / 100]));
}
var index = {
  isHex: isHex,
  isRgb: isRgb,
  isRgba: isRgba,
  isRgbOrRgba: isRgbOrRgba,
  fade: fade,
  toHex: toHex,
  toRgb: toRgb,
  darken: darken,
  lighten: lighten,
  getOpacity: getOpacity,
  getRgbValue: getRgbValue,
  getRgbaValue: getRgbaValue,
  getColorFromRgbValue: getColorFromRgbValue
};

exports.darken = darken;
exports.default = index;
exports.fade = fade;
exports.getColorFromRgbValue = getColorFromRgbValue;
exports.getOpacity = getOpacity;
exports.getRgbValue = getRgbValue;
exports.getRgbaValue = getRgbaValue;
exports.isHex = isHex;
exports.isRgb = isRgb;
exports.isRgbOrRgba = isRgbOrRgba;
exports.isRgba = isRgba;
exports.lighten = lighten;
exports.toHex = toHex;
exports.toRgb = toRgb;
